### PR TITLE
[Enhancement] (Nereids) scalar expression rewrite framework

### LINKED ISSUE #9633: [Enhancement] (Nereids) scalar expression rewrite framework
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

use visitor to add a temporary expression rewrite framework

### Solution

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by qzsee**
# Proposed changes

Issue Number: close #9633

## Problem Summary:

The scalar expression is rewritten using the `visitor pattern` as a traversal.

In the  abstract class `ExpressionVisitor`, which contains all predicate to rewrite. 

We have provided a rewrite rules interface ` ExpressionRewriteRule `, `AbstractExpressionRewriteRule` class implements the interface and expanded the `ExpressionVisitor`, if we want to realize an expression rewriting rules, Direct implementation `AbstractExpressionRewriteRule` provided in the method of traversing the predicate.

There are two rules to refer: `NormalizeExpressionRule` and `SimplifyNotExprRule`

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by morningman — COMMENTED** [review]
Please pay attention to the import order

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4 [hunk 1]
```diff
@@ -173,6 +174,10 @@ comparisonOperator
     : EQ | NEQ | LT | LTE | GT | GTE | NSEQ
     ;
 
+logicalOperator
```
Comment (morrySnow) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:177]: not use anymore?

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4 [hunk 2]
```diff
@@ -147,6 +147,7 @@ booleanExpression
 valueExpression
     : primaryExpression                                                                      #valueExpressionDefault
     | left=valueExpression comparisonOperator right=valueExpression                          #comparison
+    | NOT valueExpression                                                                    #not
```
Comment (morrySnow) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:150]: not should be placed under booleanExpression as 'NOT booleanExpression'

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4 (hunk 1 above)
Comment (qzsee) [reply fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:177]: for test . forget to delete

File: fe/fe-core/src/test/java/org/apache/doris/nereids/rules/expression/rewrite/ExpressionRewriteTest.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        assertRewrite("'a' = 1 or 'a' = 1", "'a' = 1");
+        assertRewrite("'a' = 1 or 'a' = 1 or 'b' >= 1", "'a' = 1 or 'b' >= 1");
+        */
+    }
+
+    private void assertRewrite(String expression, String expected) {
+        Expression expectedExpression = PARSER.createExpression(expected);
+        Expression needRewriteExpression = PARSER.createExpression(expression);
+        Expression rewrittenExpression = rewriter.rewrite(needRewriteExpression);
+        LOG.info("original expression : {} expected expression : {} rewritten expression : {}", needRewriteExpression, expectedExpression, rewrittenExpression);
```
Comment (jackwener) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/rules/expression/rewrite/ExpressionRewriteTest.java:80]: Remove the `LOG` in UT?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/EqualTo.java [hunk 4]
```diff
@@ -17,8 +17,13 @@
 
 package org.apache.doris.nereids.trees.expressions;
 
+import com.google.common.base.Preconditions;
 import org.apache.doris.nereids.exceptions.UnboundException;
 import org.apache.doris.nereids.trees.NodeType;
+import org.apache.doris.nereids.trees.TreeNode;
+
+import java.util.List;
+import java.util.Objects;
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/EqualTo.java:26]: redundant

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/NormalizeExpressionRule.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import org.apache.doris.nereids.trees.expressions.ComparisonPredicate;
+import org.apache.doris.nereids.trees.expressions.EqualTo;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.GreaterThan;
+import org.apache.doris.nereids.trees.expressions.GreaterThanEqual;
+import org.apache.doris.nereids.trees.expressions.LessThan;
+import org.apache.doris.nereids.trees.expressions.LessThanEqual;
+
+
+public class NormalizeExpressionRule extends AbstractExpressionRewriteRule {
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/NormalizeExpressionRule.java:33]: Rename `AdjustCmpExprOrderRule`.

normalize is so general.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/NormalizeExpressionRule.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+import org.apache.doris.nereids.trees.NodeType;
+import org.apache.doris.nereids.trees.expressions.ComparisonPredicate;
+import org.apache.doris.nereids.trees.expressions.EqualTo;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.GreaterThan;
+import org.apache.doris.nereids.trees.expressions.GreaterThanEqual;
+import org.apache.doris.nereids.trees.expressions.LessThan;
+import org.apache.doris.nereids.trees.expressions.LessThanEqual;
+
+
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/NormalizeExpressionRule.java:32]: Add example in header javadoc

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/NotExpressionRule.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.doris.nereids.trees.expressions.ComparisonPredicate;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.GreaterThan;
+import org.apache.doris.nereids.trees.expressions.GreaterThanEqual;
+import org.apache.doris.nereids.trees.expressions.LessThan;
+import org.apache.doris.nereids.trees.expressions.LessThanEqual;
+import org.apache.doris.nereids.trees.expressions.Literal;
+import org.apache.doris.nereids.trees.expressions.Not;
+
+public class NotExpressionRule extends AbstractExpressionRewriteRule {
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/NotExpressionRule.java:32]: `SimplifyNotExprRule`

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4 [hunk 8]
```diff
@@ -141,7 +141,8 @@ expression
     ;
 
 booleanExpression
-    : valueExpression                                              #predicated
+    : valueExpression                                                                        #predicated
+    | NOT booleanExpression                                                                  #not
```
Comment (morrySnow) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:145]: need at first place

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/RewriteHelper.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    public static Expression convertLiteralToSlot(Expression expr) {
+        ExpressionRewriter rewriter = new ExpressionRewriter(ConvertLiteralToSlotRule.INSTANCE);
+        return rewriter.rewrite(expr);
+    }
+
+    /**
+     * For test.
+     * convert literal to slot
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/RewriteHelper.java:42]: if just for test, it's better move it to test folder

File: fe/fe-core/src/test/java/org/apache/doris/nereids/rules/expression/rewrite/ExpressionRewriteTest.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    @Test
+    public void testNormalizeExpressionRewrite() {
+        rewriter = new ExpressionRewriter(NormalizeExpressionRule.INSTANCE);
+
+        assertRewrite("2 > x", "x < 2");
+        assertRewrite("2 >= x", "x <= 2");
+        assertRewrite("2 < x", "x > 2");
+        assertRewrite("2 <= x", "x >= 2");
+        assertRewrite("2 = x", "x = 2");
+        /*
```
Comment (morrySnow) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/rules/expression/rewrite/ExpressionRewriteTest.java:62]: remove it if unused

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/ExpressionRewriter.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+import org.apache.doris.nereids.trees.expressions.LeafExpression;
+
+import com.google.common.collect.Lists;
+
+import java.util.List;
+
+/**
+ * Expression rewrite entry, which contains all rewrite rules.
+ */
+public class ExpressionRewriter {
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/ExpressionRewriter.java:32]: I'm afraid this is duplicated with `ExpressionVisitor`.
We can traverse the expression tree in the `applyRule` or by the visitor, it's not a good design.

Comment (qzsee) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/ExpressionRewriter.java:31]: you are right . I fixed it

**Review by EmmyMiao87 — APPROVED** [review]
LGTM

**Pull request merged**